File: src/array.cpp
// src/array.cpp
#include "minml/array.h"

#include <cstdint>
#include <cstring>
#include <numeric>
#include <stdexcept>

#include "backend.h"
#include "primitives.h"

namespace minml {

namespace {

size_t product(const std::vector<size_t>& shape) {
  size_t n = 1;
  for (size_t d : shape) n *= d;
  return n;
}

std::shared_ptr<Buffer> allocate(size_t bytes, Device d) {
  switch (d) {
    case Device::CPU: return cpu_allocate(bytes);
    case Device::CUDA: return cuda_allocate(bytes);
    case Device::WebGPU: return webgpu_allocate(bytes);
  }
  throw std::runtime_error("unknown device");
}

void copy_h2d(Buffer& dst, const float* src, size_t n, Device d) {
  switch (d) {
    case Device::CPU: cpu_copy_host_to_device(dst, src, n); return;
    case Device::CUDA: cuda_copy_host_to_device(dst, src, n); return;
    case Device::WebGPU: webgpu_copy_host_to_device(dst, src, n); return;
  }
  throw std::runtime_error("unknown device");
}

void copy_d2h(const Buffer& src, float* dst, size_t n, Device d) {
  switch (d) {
    case Device::CPU: cpu_copy_device_to_host(src, dst, n); return;
    case Device::CUDA: cuda_copy_device_to_host(src, dst, n); return;
    case Device::WebGPU: webgpu_copy_device_to_host(src, dst, n); return;
  }
  throw std::runtime_error("unknown device");
}

}  // namespace

Array::Array(std::vector<float> data, std::vector<size_t> shape, Device device)
    : shape_(std::move(shape)),
      size_(product(shape_)),
      device_(device),
      dtype_(DType::Float32) {
  if (data.size() != size_)
    throw std::runtime_error("data size != product(shape)");
  data_ = allocate(size_ * dtype_bytes(dtype_), device_);
  copy_h2d(*data_, data.data(), size_, device_);
}

Array::Array(std::vector<int32_t> data, std::vector<size_t> shape, Device device)
    : shape_(std::move(shape)),
      size_(product(shape_)),
      device_(device),
      dtype_(DType::Int32) {
  if (data.size() != size_)
    throw std::runtime_error("data size != product(shape)");
  data_ = allocate(size_ * dtype_bytes(dtype_), device_);
  // copy_h2d takes float*, but the underlying memcpy on CPU treats it as
  // bytes. For non-CPU backends int32 isn't supported yet (stubs); fine.
  copy_h2d(*data_, reinterpret_cast<const float*>(data.data()), size_, device_);
}

Array::Array(std::vector<float> data, Device device)
    : shape_({data.size()}),
      size_(data.size()),
      device_(device),
      dtype_(DType::Float32) {
  data_ = allocate(size_ * dtype_bytes(dtype_), device_);
  copy_h2d(*data_, data.data(), size_, device_);
}

Array::Array(std::vector<size_t> shape, DType dtype, Device device,
             std::shared_ptr<Primitive> prim, std::vector<Array> inputs)
    : shape_(std::move(shape)),
      size_(product(shape_)),
      device_(device),
      dtype_(dtype),
      primitive_(std::move(prim)),
      inputs_(std::move(inputs)) {}

Array Array::with_batch_axis(int axis) const {
  Array out = *this;
  out.batch_axis_ = axis;
  return out;
}

Array Array::strip_batch_axis() const {
  Array out = *this;
  out.batch_axis_.reset();
  return out;
}

void Array::set_data(std::shared_ptr<Buffer> b) {
  data_ = std::move(b);
  primitive_.reset();
  inputs_.clear();
}

void Array::eval() {
  if (evaluated()) return;
  for (auto& in : inputs_) in.eval();
  data_ = allocate(size_ * dtype_bytes(dtype_), device_);
  primitive_->eval(inputs_, *this);
  primitive_.reset();
  inputs_.clear();
}

std::vector<float> Array::tolist() {
  eval();
  std::vector<float> out(size_);
  copy_d2h(*data_, out.data(), size_, device_);
  return out;
}

std::vector<int32_t> Array::tolist_int() {
  eval();
  std::vector<int32_t> out(size_);
  copy_d2h(*data_, reinterpret_cast<float*>(out.data()), size_, device_);
  return out;
}

float Array::item() {
  if (size_ != 1) throw std::runtime_error("item() requires size==1");
  return tolist()[0];
}

}  // namespace minml

File: src/array.ts
// Array — the user-facing tensor.
//
// Either evaluated (data is a Buffer) or lazy (a Primitive plus a list of
// input Arrays). Calling eval() walks the input DAG iteratively in
// post-order, allocates output buffers, and runs each primitive. The only
// async surface is tolist/tolist_int/item, which await on d2h.

import type { Buffer } from "./buffer.js";
import * as device_dispatch from "./deviceDispatch.js";
import { Device } from "./device.js";
import { DType, dtype_bytes } from "./dtype.js";
import { MinmlError } from "./error.js";
import type { Primitive } from "./primitive.js";

type State =
  | { kind: "evaluated"; buffer: Buffer }
  | { kind: "lazy"; prim: Primitive; inputs: Array[] };

// Boxed state so that all clones of the same logical Array see the same
// evaluated buffer once eval() has run on any of them. Mirrors Rust's
// Arc<Mutex<ArrayInner>>.
interface StateBox {
  state: State;
}

function product(shape: number[]): number {
  let p = 1;
  for (const d of shape) p *= d;
  return p;
}

export class Array {
  private readonly _shape: number[];
  private readonly _size: number;
  private readonly _device: Device;
  private readonly _dtype: DType;
  private readonly _batch_axis: number | null;
  private readonly _state_box: StateBox;

  private constructor(
    shape: number[],
    device: Device,
    dtype: DType,
    batch_axis: number | null,
    state_box: StateBox,
  ) {
    this._shape = shape;
    this._size = product(shape);
    this._device = device;
    this._dtype = dtype;
    this._batch_axis = batch_axis;
    this._state_box = state_box;
  }

  // ---- Eager constructors ----

  static from_f32_with_shape(
    data: Float32Array | number[],
    shape: number[],
    device: Device,
  ): Array {
    const size = product(shape);
    const arr = data instanceof Float32Array ? data : new Float32Array(data);
    if (arr.length !== size) {
      throw MinmlError.data_size(arr.length, size);
    }
    const bytes = size * dtype_bytes(DType.Float32);
    const buf = device_dispatch.allocate(device, bytes);
    device_dispatch.h2d(
      device,
      buf,
      new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength),
    );
    return new Array(shape, device, DType.Float32, null, {
      state: { kind: "evaluated", buffer: buf },
    });
  }

  static from_i32_with_shape(
    data: Int32Array | number[],
    shape: number[],
    device: Device,
  ): Array {
    const size = product(shape);
    const arr = data instanceof Int32Array ? data : new Int32Array(data);
    if (arr.length !== size) {
      throw MinmlError.data_size(arr.length, size);
    }
    const bytes = size * dtype_bytes(DType.Int32);
    const buf = device_dispatch.allocate(device, bytes);
    device_dispatch.h2d(
      device,
      buf,
      new Uint8Array(arr.buffer, arr.byteOffset, arr.byteLength),
    );
    return new Array(shape, device, DType.Int32, null, {
      state: { kind: "evaluated", buffer: buf },
    });
  }

  static from_f32_1d(data: Float32Array | number[], device: Device): Array {
    const arr = data instanceof Float32Array ? data : new Float32Array(data);
    return Array.from_f32_with_shape(arr, [arr.length], device);
  }

  static from_i32_1d(data: Int32Array | number[], device: Device): Array {
    const arr = data instanceof Int32Array ? data : new Int32Array(data);
    return Array.from_i32_with_shape(arr, [arr.length], device);
  }

  // ---- Lazy constructor ----

  static lazy(
    shape: number[],
    dtype: DType,
    device: Device,
    prim: Primitive,
    inputs: Array[],
  ): Array {
    return new Array(shape, device, dtype, null, {
      state: { kind: "lazy", prim, inputs },
    });
  }

  // ---- Accessors ----

  shape(): number[] {
    return this._shape;
  }
  size(): number {
    return this._size;
  }
  device(): Device {
    return this._device;
  }
  dtype(): DType {
    return this._dtype;
  }
  batch_axis(): number | null {
    return this._batch_axis;
  }

  evaluated(): boolean {
    return this._state_box.state.kind === "evaluated";
  }

  buffer(): Buffer | null {
    const s = this._state_box.state;
    return s.kind === "evaluated" ? s.buffer : null;
  }

  // ---- Vmap-axis tagging ----

  with_batch_axis(axis: number): Array {
    return new Array(
      this._shape,
      this._device,
      this._dtype,
      axis,
      this._state_box,
    );
  }

  strip_batch_axis(): Array {
    return new Array(
      this._shape,
      this._device,
      this._dtype,
      null,
      this._state_box,
    );
  }

  // ---- Eval ----

  // Iterative post-order DFS. No recursion. WebGPU dispatch is sync
  // (queue.submit returns immediately); only tolist/item actually await
  // on d2h.
  eval(): void {
    if (this.evaluated()) return;
    const stack: [Array, boolean][] = [[this, false]];
    while (stack.length > 0) {
      const [node, visited] = stack.pop()!;
      if (node.evaluated()) continue;
      if (!visited) {
        stack.push([node, true]);
        for (const inp of node._inputs_snapshot()) {
          if (!inp.evaluated()) stack.push([inp, false]);
        }
      } else {
        node._run_primitive();
      }
    }
  }

  private _inputs_snapshot(): Array[] {
    const s = this._state_box.state;
    return s.kind === "lazy" ? s.inputs.slice() : [];
  }

  private _run_primitive(): void {
    const s = this._state_box.state;
    if (s.kind === "evaluated") return;
    const { prim, inputs } = s;
    const bytes = this._size * dtype_bytes(this._dtype);
    const buf = device_dispatch.allocate(this._device, bytes);
    // Install on this Array first so the primitive can write through.
    this._state_box.state = { kind: "evaluated", buffer: buf };
    prim.eval(inputs, this);
  }

  // ---- Async readback ----

  async tolist(): Promise<number[]> {
    this.eval();
    const out = new Float32Array(this._size);
    const buf = this.buffer()!;
    await device_dispatch.d2h_async(
      this._device,
      buf,
      new Uint8Array(out.buffer, 0, out.byteLength),
    );
    return [...out];
  }

  async tolist_int(): Promise<number[]> {
    this.eval();
    const out = new Int32Array(this._size);
    const buf = this.buffer()!;
    await device_dispatch.d2h_async(
      this._device,
      buf,
      new Uint8Array(out.buffer, 0, out.byteLength),
    );
    return [...out];
  }

  async item(): Promise<number> {
    if (this._size !== 1) throw MinmlError.item_requires_size_1();
    return (await this.tolist())[0];
  }
}
